### ISSUE TITLE
`noUncheckedIndexedAccess`: unexpected error when indexing with generic constrained to `keyof`

### ISSUE THREAD

**Issue by OliverJAsh** [microsoft/TypeScript#40666]
In the process of enabling the new `noUncheckedIndexedAccess` rule on my project, I came across this interesting error:

```ts
type MyRecord = { a: number; b: string };

declare const myRecord: MyRecord;

declare const key: 'a' | 'b';
const value = myRecord[key]; // string | number ✅

// ❌ Unexpected error
// Type 'MyRecord[Key] | undefined' is not assignable to type 'MyRecord[Key]'
const fn = <Key extends keyof MyRecord>(key: Key): MyRecord[Key] => myRecord[key];
```

In this case I did not expect `myRecord[key]` to return type `MyRecord[Key] | undefined`, because the `key` is constrained to `keyof MyRecord`.

_Originally posted by @OliverJAsh in https://github.com/microsoft/TypeScript/issues/13778#issuecomment-696024937_
